fabric-input-events-v1/src/main/java/net/fabricmc/fabric/api/event/client/input/CharEvent.java

```diff
@@ -0,0 +1,35 @@
+/*
+ * Copyright (c) 2016, 2017, 2018, 2019 FabricMC
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package net.fabricmc.fabric.api.event.client.input;
+
+public class CharEvent {
```

Comment:
From the big guy himself (modmuss)

These classes need to be inlined into the event interfaces. I don't see any real reason to expose any interfaces here as there is very little extra stuff you could be exposing here.

Simply inline the methods and all into the parameters of each callback interface.

Reply:
For this event, definitely, but for `(double x, double y, double dx, double dy, int buttons, int modKeys, double scrollX, double scrollY) -> { /* ... */ }` and `(int code, int scancode, int action, int modKeys, Key key, Keybind keybind) -> { /* ... */ }`? My intent is to move away from multi-parameter event listeners entirely since they have several drawbacks and no direct benefits but we'll save that for another pull request. I will try to remove the object allocations at the very least.